<a href="https://colab.research.google.com/github/Blackman9t/EDA/blob/master/foursquare_api_exploring_lagos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Introduction
In this notebook, we'd make calls to the Foursquare API for different purposes. to explore a particular venue{Radisson Blu Anchorage Hotel),  to explore a Foursquare user who has made a review on Radisson Blu Hotel l to explore the geographical location, and to get trending venues around Radisson Blu Hotel. Also, we will use the visualization library, Folium, to visualize the results.

import needed libraries

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

#!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Folium installed
Libraries imported.


Define Foursquare credentials

In [98]:
CLIENT_ID = 'L2XUYCRD44VP2YBDMZ3XY0Z4XKO1JODM5SMG3IIYRZ5330FD' # your Foursquare ID
CLIENT_SECRET = '5KI00FG4RT2VJQTIDPXFY1N4AOKTIXLIZEAO0AHPP3DPD0U1' # your Foursquare Secret
VERSION = '20190620'  # Version should be todays date!
LIMIT = 100
print('Your credentials:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentials:
CLIENT_ID: L2XUYCRD44VP2YBDMZ3XY0Z4XKO1JODM5SMG3IIYRZ5330FD
CLIENT_SECRET:5KI00FG4RT2VJQTIDPXFY1N4AOKTIXLIZEAO0AHPP3DPD0U1


Let's again assume that you are staying at the Eko hotel. So let's start by converting the Eko Hotel's address to its latitude and longitude coordinates.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent foursquare_agent, as shown below.

In [3]:
address = '1 Ozumba Mbadiwe Avenue, Victoria Island Lagos, NG'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

print(latitude, longitude)


6.43642025 3.41639455


In [4]:
type(location)

# location is a location object of geopy class

geopy.location.Location

Now, let's assume that it is lunch time, and you are craving Pizza. So, let's define a query to search for Pizza restaurants that is within 1500 metres(1.5 km) from the Radisson Blu.

In [5]:
search_query = 'Pizza'
radius = 1500
print(search_query + ' .... OK!')

Pizza .... OK!


Let's define the proper URI to return the Pizza restaurants within 1500 meters of Eko-Hotel

In [6]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'\
.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=L2XUYCRD44VP2YBDMZ3XY0Z4XKO1JODM5SMG3IIYRZ5330FD&client_secret=5KI00FG4RT2VJQTIDPXFY1N4AOKTIXLIZEAO0AHPP3DPD0U1&ll=6.43642025,3.41639455&v=20190620&query=Pizza&radius=1500&limit=30'

Send the get query and analyse results

In [7]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d0bff5d23bb8e0035741e46'},
 'response': {'venues': [{'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
       'suffix': '.png'},
      'id': '4bf58dd8d48988d1ca941735',
      'name': 'Pizza Place',
      'pluralName': 'Pizza Places',
      'primary': True,
      'shortName': 'Pizza'}],
    'hasPerk': False,
    'id': '4f805d88e4b05766b1888536',
    'location': {'address': "13 Musa Yar'Adua",
     'cc': 'NG',
     'city': 'Victoria Island',
     'country': 'Nigeria',
     'crossStreet': 'Adeleke Adedoyin Street',
     'distance': 450,
     'formattedAddress': ["13 Musa Yar'Adua (Adeleke Adedoyin Street)",
      'Victoria Island',
      'Lagos',
      'Nigeria'],
     'labeledLatLngs': [{'label': 'display',
       'lat': 6.432408,
       'lng': 3.416935}],
     'lat': 6.432408,
     'lng': 3.416935,
     'neighborhood': 'Victoria Island',
     'state': 'Lagos'},
    'name': 'Pizze-riah Wood-fired pizza',
    're

Get relevant part of JSON and transform it into a pandas dataframe


In [8]:
# assign relevant part of JSON to venues
venues = results['response']['venues']
venues

[{'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
     'suffix': '.png'},
    'id': '4bf58dd8d48988d1ca941735',
    'name': 'Pizza Place',
    'pluralName': 'Pizza Places',
    'primary': True,
    'shortName': 'Pizza'}],
  'hasPerk': False,
  'id': '4f805d88e4b05766b1888536',
  'location': {'address': "13 Musa Yar'Adua",
   'cc': 'NG',
   'city': 'Victoria Island',
   'country': 'Nigeria',
   'crossStreet': 'Adeleke Adedoyin Street',
   'distance': 450,
   'formattedAddress': ["13 Musa Yar'Adua (Adeleke Adedoyin Street)",
    'Victoria Island',
    'Lagos',
    'Nigeria'],
   'labeledLatLngs': [{'label': 'display', 'lat': 6.432408, 'lng': 3.416935}],
   'lat': 6.432408,
   'lng': 3.416935,
   'neighborhood': 'Victoria Island',
   'state': 'Lagos'},
  'name': 'Pizze-riah Wood-fired pizza',
  'referralId': 'v-1561067357'},
 {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
     'suffix': '.png'},
    'id': 

In [9]:
# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

categories  ...    referralId
0  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  ...  v-1561067357
1  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  ...  v-1561067357
2  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  ...  v-1561067357
3  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  ...  v-1561067357
4  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  ...  v-1561067357

[5 rows x 17 columns]

In [10]:
dataframe.shape

(10, 17)

Define information of interest and filter dataframe


In [11]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name   categories  ...  state                        id
0  Pizze-riah Wood-fired pizza  Pizza Place  ...  Lagos  4f805d88e4b05766b1888536
1               Domino's Pizza  Pizza Place  ...  Lagos  5041105be4b0af4f728b1946
2                    Pizza Hut  Pizza Place  ...  Lagos  5c4230419deb7d002cc18980
3               Domino's Pizza  Pizza Place  ...  Lagos  5027d6aee4b0c80453509155
4                     La Pizza  Pizza Place  ...  Lagos  519e17c1498e8aa36af27ab0
5             Debonair's Pizza  Pizza Place  ...  Lagos  58372a9ca6118c60b52d5715
6                Pizza Marella   Restaurant  ...    NaN  53a80e0b498e1f19d246e780
7              Debonairs pizza  Pizza Place  ...  Lagos  4c6275ac58810f470c32071e
8                Dominos Pizza  Pizza Place  ...  Lagos  5b8fff0812c8f00039e5cc01
9                Dominos Pizza         None  ...    NaN  518bf3b7e4b0dabd6ec192a5

[10 rows x 15 columns]

In [12]:
dataframe_filtered.shape

(10, 15)

Let's visualize the Pizza restaurants that are nearby

In [13]:
dataframe_filtered.name

0    Pizze-riah Wood-fired pizza
1                 Domino's Pizza
2                      Pizza Hut
3                 Domino's Pizza
4                       La Pizza
5               Debonair's Pizza
6                  Pizza Marella
7                Debonairs pizza
8                  Dominos Pizza
9                  Dominos Pizza
Name: name, dtype: object

In [14]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Eko Hotel

# add a red circle marker to represent the Eko Hotel
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Eko Hotel',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the Chinese restaurants as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

### Let's explore the closest Pizza restaurant -- Dominos Pizza


In [15]:
venue_id = '5041105be4b0af4f728b1946' # ID of Dominos Pizza
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v2/venues/5041105be4b0af4f728b1946?client_id=L2XUYCRD44VP2YBDMZ3XY0Z4XKO1JODM5SMG3IIYRZ5330FD&client_secret=5KI00FG4RT2VJQTIDPXFY1N4AOKTIXLIZEAO0AHPP3DPD0U1&v=20190620'

In [16]:
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'price', 'likes', 'dislike', 'ok', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'pageUpdates', 'inbox', 'attributes'])


{'allowMenuUrlEdit': True,
 'attributes': {'groups': [{'count': 1,
    'items': [{'displayName': 'Price', 'displayValue': '$', 'priceTier': 1}],
    'name': 'Price',
    'summary': '$',
    'type': 'price'}]},
 'beenHere': {'count': 0,
  'lastCheckinExpiredAt': 0,
  'marked': False,
  'unconfirmedCount': 0},
 'canonicalUrl': 'https://foursquare.com/v/dominos-pizza/5041105be4b0af4f728b1946',
 'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
    'suffix': '.png'},
   'id': '4bf58dd8d48988d1ca941735',
   'name': 'Pizza Place',
   'pluralName': 'Pizza Places',
   'primary': True,
   'shortName': 'Pizza'}],
 'contact': {},
 'createdAt': 1346441307,
 'dislike': False,
 'hereNow': {'count': 0, 'groups': [], 'summary': 'Nobody here'},
 'id': '5041105be4b0af4f728b1946',
 'inbox': {'count': 0, 'items': []},
 'likes': {'count': 2,
  'groups': [{'count': 2,
    'items': [{'firstName': 'Oluwayemisola',
      'gender': 'female',
      'id': '91977840',
      '

### Get the venue's rating for the closest restaurant- Dominos Pizza

In [17]:
closest_venue_id = '5041105be4b0af4f728b1946' # ID Dominos Pizza

url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['likes']['count'])
except:
    print('This venue has not been rated yet.')

2


That is not a very good rating. As it has only two likes.  Let's check the rating of all the Pizza restaurants and return the name and rating of the highest rated restaurant

In [18]:
def find_highest_rated(dataframe_filtered):
    id_list = list(dataframe_filtered.id)
    id_name = list(dataframe_filtered.name)
    max_rating = 0
    name = ''
    
    for i,j in zip(id_list, id_name):
        url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(i, CLIENT_ID, CLIENT_SECRET, VERSION)
        result = requests.get(url).json()
        try:
            x = result['response']['venue']['likes']['count']
            if x > max_rating:
                max_rating = x
                name = j
                
        except:
            pass
    return (max_rating, name)
   

max_rated_restaurant = find_highest_rated(dataframe_filtered)
max_rated_restaurant

(36, 'Debonairs pizza')

Let's see some of Debonairs Pizza Stats

In [19]:
best_venue_id = '4c6275ac58810f470c32071e' # ID Debonairs Pizza

url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(best_venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['likes'])
except:
    print('This venue has not been rated yet.')

{'count': 36, 'groups': [{'type': 'others', 'count': 36, 'items': []}], 'summary': '36 Likes'}


Debonairs Pizza has the highest Rating as at today 20th June 2019

Let's see some of the likes for Debonairs pizza

In [20]:
best_venue_id = '4c6275ac58810f470c32071e' # ID Debonairs Pizza

url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(best_venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

best_result = requests.get(url).json()
print(result['response']['venue'].keys())

best_result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'price', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'popular', 'pageUpdates', 'inbox', 'parent', 'hierarchy', 'attributes', 'bestPhoto', 'colors'])


{'allowMenuUrlEdit': True,
 'attributes': {'groups': [{'count': 1,
    'items': [{'displayName': 'Price', 'displayValue': '$', 'priceTier': 1}],
    'name': 'Price',
    'summary': '$',
    'type': 'price'},
   {'count': 3,
    'items': [{'displayName': 'Reservations', 'displayValue': 'No'}],
    'name': 'Reservations',
    'type': 'reservations'},
   {'count': 5,
    'items': [{'displayName': 'Credit Cards', 'displayValue': 'Yes'}],
    'name': 'Credit Cards',
    'summary': 'Credit Cards',
    'type': 'payments'},
   {'count': 1,
    'items': [{'displayName': 'Outdoor Seating', 'displayValue': 'Yes'}],
    'name': 'Outdoor Seating',
    'summary': 'Outdoor Seating',
    'type': 'outdoorSeating'},
   {'count': 1,
    'items': [{'displayName': 'Wi-Fi', 'displayValue': 'Yes'}],
    'name': 'Wi-Fi',
    'summary': 'Wi-Fi',
    'type': 'wifi'},
   {'count': 8,
    'items': [{'displayName': 'Brunch', 'displayValue': 'Brunch'},
     {'displayName': 'Lunch', 'displayValue': 'Lunch'},
     {'

### Let's get the venue tips<br>
Create URL and send GET request. Make sure to set limit to get all tips

In [21]:
## Debonairs tips
limit = 15 # set limit to be greater than or equal to the total number of tips
url = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(best_venue_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

best_results = requests.get(url).json()
best_results['response'].keys()

dict_keys(['tips'])

Get tips and list of associated features...<br>
Remember we only get two tips since this is a personal acct

In [22]:
tips = best_results['response']['tips']['items']
tips

[{'agreeCount': 1,
  'canonicalUrl': 'https://foursquare.com/item/5458b704498e1ab64b727436',
  'createdAt': 1415100164,
  'disagreeCount': 0,
  'id': '5458b704498e1ab64b727436',
  'lang': 'en',
  'likes': {'count': 1,
   'groups': [{'count': 1,
     'items': [{'firstName': 'Lucky',
       'gender': 'male',
       'id': '119367471',
       'lastName': 'D',
       'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
        'suffix': '/119367471-3BN0J3RIPBZ431YN.jpg'}}],
     'type': 'others'}],
   'summary': '1 like'},
  'logView': True,
  'text': 'The triple Decker anything is a treat you should enjoy as often as you can.',
  'todo': {'count': 1},
  'type': 'user',
  'user': {'firstName': 'Omobola',
   'gender': 'female',
   'id': '94019944',
   'lastName': 'F',
   'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
    'suffix': '/94019944-J3UVYYZKSG31TMXK.jpg'}}},
 {'agreeCount': 0,
  'authorInteractionType': 'liked',
  'canonicalUrl': 'https://foursquare.com/item/5725350a49

In [23]:
tip = best_results['response']['tips']['items'][0]
tip.keys()

dict_keys(['id', 'createdAt', 'text', 'type', 'canonicalUrl', 'lang', 'likes', 'logView', 'agreeCount', 'disagreeCount', 'todo', 'user'])

Format column width and display all tips

In [24]:
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips) # converting a json file to a pandas dataframe

# columns to keep
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.gender', 'user.id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display tips
tips_filtered

text  ...   user.id
0  The triple Decker anything is a treat you should enjoy as often as you can.  ...  94019944
1  Everything!!!! Especially the meaty triple decker pizza... amazing much      ...  28338092

[2 rows x 8 columns]

### Search a Foursquare User who has rated radisson Blu hotel before

In [25]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}'\
.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION)
url

'https://api.foursquare.com/v2/venues/search?client_id=L2XUYCRD44VP2YBDMZ3XY0Z4XKO1JODM5SMG3IIYRZ5330FD&client_secret=5KI00FG4RT2VJQTIDPXFY1N4AOKTIXLIZEAO0AHPP3DPD0U1&ll=6.43642025,3.41639455&v=20190620'

In [26]:
results = requests.get(url).json()
results['response'].keys()

dict_keys(['venues', 'confident'])

In [28]:
venues = results['response']['venues']
dataframe = json_normalize(venues)
dataframe.head()

categories  ...    referralId
0  [{'id': '4bf58dd8d48988d1c9941735', 'name': 'Ice Cream Shop', 'pluralName': 'Ice Cream Shops', 'shortName': 'Ice Cream', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/icecream_', 'suffix': '.png'}, 'primary': True}]                                ...  v-1561067365
1  [{'id': '4bf58dd8d48988d1fa931735', 'name': 'Hotel', 'pluralName': 'Hotels', 'shortName': 'Hotel', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/travel/hotel_', 'suffix': '.png'}, 'primary': True}]                                                       ...  v-1561067365
2  [{'id': '4bf58dd8d48988d176941735', 'name': 'Gym', 'pluralName': 'Gyms', 'shortName': 'Gym', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/gym_', 'suffix': '.png'}, 'primary': True}]                                                             ...  v-1561067365
3  [{'id': '4bf58dd8d48988d124941735', 'name': 'Office', 'pluralName': 'Offices', 'shortName': 'Office', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/default_', 'suffix': '.png'}, 'primary': True}]                                                ...  v-1561067365
4  [{'id': '4bf58dd8d48988d1f1931735', 'name': 'General Entertainment', 'pluralName': 'General Entertainment', 'shortName': 'Entertainment', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/arts_entertainment/default_', 'suffix': '.png'}, 'primary': True}]  ...  v-1561067365

[5 rows x 16 columns]

In [32]:
radisson_blu_id = str(dataframe.at[1, 'id'])
radisson_blu_id

'4d9325d3b32737042aad9125'

In [34]:
radisson_id = radisson_blu_id # ID Radisson blu hotel

url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(radisson_id, CLIENT_ID, CLIENT_SECRET, VERSION)

radisson_result = requests.get(url).json()
print(radisson_result['response']['venue'].keys())

radisson_result['response']['venue']['likes']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'url', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'description', 'storeId', 'page', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'pageUpdates', 'inbox', 'attributes', 'bestPhoto', 'colors'])


{'count': 118,
 'groups': [{'count': 118, 'items': [], 'type': 'others'}],
 'summary': '118 Likes'}

We can see radisson blu as at today(20160620) has 118 likes already

In [54]:
radisson_tips = radisson_result['response']['venue']['tips']['groups']
radisson_tips

[{'count': 44,
  'items': [{'agreeCount': 20,
    'canonicalUrl': 'https://foursquare.com/item/534bf9d4498e7b69ab14bbad',
    'createdAt': 1397488084,
    'disagreeCount': 0,
    'id': '534bf9d4498e7b69ab14bbad',
    'lang': 'en',
    'likes': {'count': 20,
     'groups': [{'count': 20, 'items': [], 'type': 'others'}],
     'summary': '20 likes'},
    'logView': True,
    'photo': {'createdAt': 1397488089,
     'height': 720,
     'id': '534bf9d911d25a0cbed30e13',
     'prefix': 'https://fastly.4sqi.net/img/general/',
     'source': {'name': 'Foursquare Web', 'url': 'https://foursquare.com'},
     'suffix': '/2960181_as8Yt92x2EDWQF9wn15PK4fF-bwU7E-rh2vGWKp1OqA.jpg',
     'visibility': 'public',
     'width': 960},
    'photourl': 'https://fastly.4sqi.net/img/general/original/2960181_as8Yt92x2EDWQF9wn15PK4fF-bwU7E-rh2vGWKp1OqA.jpg',
    'text': 'The View bar and lounge is the perfect place to sip a cool cocktail and enjoy the view across the lagoon of Lagos. More about Lagos in the LH M

In [0]:
# dataframe = json_normalize(radisson_tips) # flatten JSON

# # filter columns
# filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
# dataframe_filtered = dataframe.loc[:, filtered_columns]

# # filter the category for each row
# dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# # clean columns
# dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

# dataframe_filtered.head(10)

## 3. Search a Foursquare User
> `https://api.foursquare.com/v2/users/`**USER_ID**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**

In [87]:
user_id = '12610843' # user ID with most agree counts and complete profile
# 94019944
#12610843

url = 'https://api.foursquare.com/v2/users/{}?client_id={}&client_secret={}&v={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION) # define URL

# send GET request
results = requests.get(url).json()
user_data = results['response']['user']

# display features associated with user
user_data.keys()

dict_keys(['id', 'firstName', 'lastName', 'gender', 'canonicalUrl', 'photo', 'friends', 'tips', 'homeCity', 'bio', 'contact', 'photos', 'type', 'mayorships', 'checkins', 'lists', 'lenses'])

In [88]:
print('First Name: ' + user_data['firstName'])
print('Last Name: ' + user_data['lastName'])
print('Gender: ' + user_data['gender'])
print('Home City: ' + user_data['homeCity'])

First Name: Tayo
Last Name: A
Gender: female
Home City: Lagos, Nigeria


How many tips has this user submitted?

In [89]:
user_data['tips']

{'count': 18}

Get User's tips

In [90]:
# define tips URL
url = 'https://api.foursquare.com/v2/users/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

# send GET request and get user's tips
user_results = requests.get(url).json()
tips = user_results['response']['tips']['items']

# format column width
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips)

# filter columns
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display user's tips
tips_filtered

text  ...                        id
0   Try the tennis n the banga                                                                                                                                                       ...  523b5cfc11d25e24e91f43fa
1   Shisha in different flavours for just N1000                                                                                                                                      ...  51ce0e89498e2c23f4c05b36
2   Fish & Chips is divine                                                                                                                                                           ...  51c711de498e86992720134e
3   Try Beef Burger @ d ABIBIZ resturant                                                                                                                                             ...  51c1c3a6498e60fa96ba3b02
4   Nice couldn't be better                                                                                                                                                          ...  4f44a403e4b025d84374e0dd
5   Amazingly gorgeous                                                                                                                                                               ...  4f10700de4b09e81d794184f
6   Do you know Raddisson blu is Supporting GB&Godyva's costume Party go get your tickets now cos there is a free One night booking plus meal worth 75k for the best dressed couple  ...  4ece4321f5b9832ad7c1eb1b
7   Its America in Nigeria, love it here, the locals r cool as well.                                                                                                                 ...  4e7bebab775b1336d3e59d78
8   This street links about 10 major roads really great 4 avoiding traffic                                                                                                           ...  4e632035b0fb188e8e022ee9
9   Wed night is buffet                                                                                                                                                              ...  4e5fe90d2271573ad6811c6f
10  Primark is right opposite                                                                                                                                                        ...  4e50461dd164978659d72b72
11  8yrs I pass here daily in cov.                                                                                                                                                   ...  4e50460962e14b77e3a9144f
12  Hmmmmmm D view, serene, ambience its on point.                                                                                                                                   ...  4e503a89ae6051db8dd86866
13  Its not so great but hey! Nigeria is a good Country, I believe we will get there. My fav airport in d world Schiphol Amsterdam                                                   ...  4e503a32ae6051db8dd862b2
14  Absolutely Spot on                                                                                                                                                               ...  4e5038821495fc2ca40b0e33

[15 rows x 4 columns]

Let's get the venue for the tip with the greatest number of agree counts

In [91]:
tip_id = '51ce0e89498e2c23f4c05b36' # tip id

# define URL
url = 'http://api.foursquare.com/v2/tips/{}?client_id={}&client_secret={}&v={}'.format(tip_id, CLIENT_ID, CLIENT_SECRET, VERSION)

# send GET Request and examine results
result = requests.get(url).json()
print(result['response']['tip']['venue']['name'])
print(result['response']['tip']['venue']['location'])

Pool Golden Tulip
{'lat': 6.4652548652239545, 'lng': 3.3025544097330877, 'labeledLatLngs': [{'label': 'display', 'lat': 6.4652548652239545, 'lng': 3.3025544097330877}], 'cc': 'NG', 'city': 'Lagos', 'state': 'Lagos', 'country': 'Nigeria', 'formattedAddress': ['Lagos', 'Lagos', 'Nigeria']}


Get users Friends

In [92]:
user_friends = json_normalize(user_data['friends']['groups'][0]['items'])
user_friends

bio firstName  ...                    photo.suffix tips.count
0                   Olla      ...  /EFWGPLYB0GKNEC3S.jpg           2        
1                   Sunday    ...  /80621785-5EEF404YOKLD0VF3.jpg  1        
2  Always on point  Onyeka    ...  /91427432-2521BX1GVWTBSN1N.jpg  1        
3                   Cheri     ...  /8517926-C0Z24WE11FRC1HV0.jpg   25       
4                   Kollins   ...  /87099052-VZMRDPM43HCWNNRU.jpg  0        
5  Adventurer!      John      ...  /3387004-GVQXBYTR1S1Z055I.jpg   3        
6  Hard work pays   Sotonye   ...  /76988229-CW2MOUFP4BMYGMCU.jpg  10       
7  2BB117D4         Richyvic  ...  /88784972-CEGZXTNWUWIDPCTZ.jpg  0        
8                   Oyegade   ...  /blank_boy.png                  0        
9                   Deere     ...  /blank_boy.png                  0        

[10 rows x 11 columns]

### Retrieve the User's Profile Image


In [93]:
user_data

{'bio': '',
 'canonicalUrl': 'https://foursquare.com/dredsigirl',
 'checkins': {'count': 197, 'items': []},
 'contact': {},
 'firstName': 'Tayo',
 'friends': {'count': 137,
  'groups': [{'count': 137,
    'items': [{'bio': '',
      'contact': {},
      'firstName': 'Olla',
      'gender': 'female',
      'homeCity': 'PORT HARCOURT',
      'id': '2630915',
      'lastName': 'K',
      'lists': {'groups': [{'count': 2, 'items': [], 'type': 'created'}]},
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/EFWGPLYB0GKNEC3S.jpg'},
      'tips': {'count': 2}},
     {'bio': '',
      'contact': {},
      'firstName': 'Sunday',
      'gender': 'male',
      'homeCity': '',
      'id': '80621785',
      'lastName': 'A',
      'lists': {'groups': [{'count': 2, 'items': [], 'type': 'created'}]},
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/80621785-5EEF404YOKLD0VF3.jpg'},
      'tips': {'count': 1}},
     {'bio': 'Always on point

In [94]:
# 1. grab prefix of photo
# 2. grab suffix of photo
# 3. concatenate them using the image size  
Image(url='https://igx.4sqi.net/img/user/300x300/8517926-C0Z24WE11FRC1HV0.jpg')
#Image(url='https://fastly.4sqi.net/img/user/300x300/94019944-J3UVYYZKSG31TMXK.jpg')

## Explore a location
> `https://api.foursquare.com/v2/venues/`**explore**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**`&limit=`**LIMIT**

In [100]:
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=L2XUYCRD44VP2YBDMZ3XY0Z4XKO1JODM5SMG3IIYRZ5330FD&client_secret=5KI00FG4RT2VJQTIDPXFY1N4AOKTIXLIZEAO0AHPP3DPD0U1&ll=6.43642025,3.41639455&v=20190620&radius=1500&limit=100'

In [101]:
results = requests.get(url).json()
'There are {} around Radisson Blu Hotel.'.format(len(results['response']['groups'][0]['items']))

'There are 100 around Radisson Blu Hotel.'

In [103]:
items = results['response']['groups'][0]['items']
items[0]

{'reasons': {'count': 0,
  'items': [{'reasonName': 'globalInteractionReason',
    'summary': 'This spot is popular',
    'type': 'general'}]},
 'referralId': 'e-0-556b4c01498e7fd1525edbab-0',
 'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/icecream_',
     'suffix': '.png'},
    'id': '4bf58dd8d48988d1c9941735',
    'name': 'Ice Cream Shop',
    'pluralName': 'Ice Cream Shops',
    'primary': True,
    'shortName': 'Ice Cream'}],
  'id': '556b4c01498e7fd1525edbab',
  'location': {'address': 'Radisson Blu',
   'cc': 'NG',
   'city': 'Victoria Island',
   'country': 'Nigeria',
   'crossStreet': 'Ozumba Mbadiwe',
   'distance': 16,
   'formattedAddress': ['Radisson Blu (Ozumba Mbadiwe)',
    'Victoria Island',
    'Lagos',
    'Nigeria'],
   'labeledLatLngs': [{'label': 'display',
     'lat': 6.436521550578695,
     'lng': 3.416506506247353}],
   'lat': 6.436521550578695,
   'lng': 3.416506506247353,
   'state': 'Lagos'},
  'name': 'Hans & René'

###Process JSON and convert it to a clean dataframe

In [112]:
dataframe = json_normalize(items) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# filter the category for each row
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean columns
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head(10)

name  ...                        id
0  Hans & René                   ...  556b4c01498e7fd1525edbab
1  Radisson Blu Anchorage Hotel  ...  4d9325d3b32737042aad9125
2  The Backyard Bar & Grill      ...  593c79d73362735a7e142ab9
3  Crust & Cream                 ...  543155d4498e648f00ac0c38
4  Churrasco                     ...  4be1e1e44283c9b612cd54f8
5  Club 57                       ...  5498b172498e81619ce8f6b4
6  Pizze-riah Wood-fired pizza   ...  4f805d88e4b05766b1888536
7  Craft Gourmet                 ...  56caf22d498e9ce2cbd69ad0
8  Lagoon Restaurant             ...  4be446d82457a593fefba915
9  The homestores (Art Cafe)     ...  530c7fc4498ee69cc24966f5

[10 rows x 16 columns]

In [113]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


# add radisson Blu as a red circle mark
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
    popup='Ecco',
    fill=True,
    color='red',
    fill_color='red',
    fill_opacity=0.6
    ).add_to(venues_map)


# add popular spots to the map as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        fill=True,
        color='blue',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(venues_map)

# display map
venues_map

## Explore Trending Venues
> `https://api.foursquare.com/v2/venues/`**trending**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**

Now, instead of simply exploring the area around Radisson Blu, you are interested in knowing the venues that are trending at the time you are done with your lunch, meaning the places with the highest foot traffic. So let's do that and get the trending venues around Radisson Blu.

In [114]:
# define URL
url = 'https://api.foursquare.com/v2/venues/trending?client_id={}&client_secret={}&ll={},{}&v={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION)

# send GET request and get trending venues
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d0c12436adbf5002c03e130'},
 'response': {'venues': []}}

In [0]:
if len(results['response']['venues']) == 0:
    trending_venues_df = 'No trending venues are available at the moment!'
    
else:
    trending_venues = results['response']['venues']
    trending_venues_df = json_normalize(trending_venues)

    # filter columns
    columns_filtered = ['name', 'categories'] + ['location.distance', 'location.city', 'location.postalCode', 'location.state', 'location.country', 'location.lat', 'location.lng']
    trending_venues_df = trending_venues_df.loc[:, columns_filtered]

    # filter the category for each row
    trending_venues_df['categories'] = trending_venues_df.apply(get_category_type, axis=1)

In [116]:
# display trending venues
trending_venues_df

'No trending venues are available at the moment!'

### Visualize Trending Values

In [0]:
if len(results['response']['venues']) == 0:
    venues_map = 'Cannot generate visual as no trending venues are available at the moment!'

else:
    venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


    # add Ecco as a red circle mark
    folium.CircleMarker(
        [latitude, longitude],
        radius=10,
        popup='Ecco',
        fill=True,
        color='red',
        fill_color='red',
        fill_opacity=0.6
    ).add_to(venues_map)


    # add the trending venues as blue circle markers
    for lat, lng, label in zip(trending_venues_df['location.lat'], trending_venues_df['location.lng'], trending_venues_df['name']):
        folium.CircleMarker(
            [lat, lng],
            radius=5,
            poup=label,
            fill=True,
            color='blue',
            fill_color='blue',
            fill_opacity=0.6
        ).add_to(venues_map)

In [118]:
# display map
venues_map

'Cannot generate visual as no trending venues are available at the moment!'